File: web/lib/preprocess.ts
/**
 * Browser-side reimplementation of src/neurolink/data/preprocess.py.
 *
 * This file must mirror the Python pipeline step for step. If the two drift, the
 * model receives inputs unlike anything it trained on and the live demo silently
 * lies. That risk is why the site ships each sample's raw JPG alongside the
 * PyTorch probabilities computed on the server: lib/verify.ts re-runs this
 * pipeline on the raw file and checks the browser reproduces the Python result.
 *
 * Pipeline: grayscale -> square resize (undo 2x horizontal stretch) -> head mask
 * -> square bbox crop with margin -> percentile clip inside mask -> resize ->
 * normalise -> 3 channels.
 */

export interface Gray {
  data: Float32Array; // 0..255
  w: number;
  h: number;
}

/** OpenCV's getGaussianKernel(5, sigma=0) => sigma 1.1. Hardcoded to match exactly. */
const GAUSS5 = [0.07076984, 0.24446331, 0.3695337, 0.24446331, 0.07076984];

export function toGray(img: ImageData): Gray {
  const { width: w, height: h, data } = img;
  const out = new Float32Array(w * h);
  for (let i = 0, p = 0; i < data.length; i += 4, p++) {
    // Channels are identical in this dataset; use OpenCV's BGR2GRAY weights
    // so any non-grayscale upload degrades the same way Python would.
    out[p] = 0.299 * data[i] + 0.587 * data[i + 1] + 0.114 * data[i + 2];
  }
  return { data: out, w, h };
}

/** Bilinear resample. Matches cv2.INTER_LINEAR closely enough for our tolerance. */
export function resize(src: Gray, nw: number, nh: number): Gray {
  const out = new Float32Array(nw * nh);
  const sx = src.w / nw;
  const sy = src.h / nh;
  for (let y = 0; y < nh; y++) {
    const fy = Math.min(src.h - 1, Math.max(0, (y + 0.5) * sy - 0.5));
    const y0 = Math.floor(fy);
    const y1 = Math.min(y0 + 1, src.h - 1);
    const wy = fy - y0;
    for (let x = 0; x < nw; x++) {
      const fx = Math.min(src.w - 1, Math.max(0, (x + 0.5) * sx - 0.5));
      const x0 = Math.floor(fx);
      const x1 = Math.min(x0 + 1, src.w - 1);
      const wx = fx - x0;
      const a = src.data[y0 * src.w + x0];
      const b = src.data[y0 * src.w + x1];
      const c = src.data[y1 * src.w + x0];
      const d = src.data[y1 * src.w + x1];
      out[y * nw + x] = a * (1 - wx) * (1 - wy) + b * wx * (1 - wy) + c * (1 - wx) * wy + d * wx * wy;
    }
  }
  return { data: out, w: nw, h: nh };
}

function blur5(src: Gray): Gray {
  const { w, h } = src;
  const tmp = new Float32Array(w * h);
  const out = new Float32Array(w * h);
  const clamp = (v: number, lo: number, hi: number) => (v < lo ? lo : v > hi ? hi : v);
  for (let y = 0; y < h; y++)
    for (let x = 0; x < w; x++) {
      let s = 0;
      for (let k = -2; k <= 2; k++) s += GAUSS5[k + 2] * src.data[y * w + clamp(x + k, 0, w - 1)];
      tmp[y * w + x] = s;
    }
  for (let y = 0; y < h; y++)
    for (let x = 0; x < w; x++) {
      let s = 0;
      for (let k = -2; k <= 2; k++) s += GAUSS5[k + 2] * tmp[clamp(y + k, 0, h - 1) * w + x];
      out[y * w + x] = s;
    }
  return { data: out, w, h };
}

/** Percentile of a 0..255 array via a 256-bin histogram (exact for 8-bit data). */
export function percentile(vals: Float32Array, idx: Uint8Array | null, p: number): number {
  const hist = new Uint32Array(256);
  let n = 0;
  for (let i = 0; i < vals.length; i++) {
    if (idx && !idx[i]) continue;
    hist[Math.max(0, Math.min(255, Math.round(vals[i])))]++;
    n++;
  }
  if (!n) return 0;
  const target = (p / 100) * (n - 1);
  let cum = 0;
  for (let b = 0; b < 256; b++) {
    cum += hist[b];
    if (cum > target) return b;
  }
  return 255;
}

/** Separable max/min filters -- a square structuring element factorises. */
function morph(mask: Uint8Array, w: number, h: number, r: number, dilate: boolean): Uint8Array {
  const pick = dilate ? Math.max : Math.min;
  const tmp = new Uint8Array(w * h);
  const out = new Uint8Array(w * h);
  for (let y = 0; y < h; y++)
    for (let x = 0; x < w; x++) {
      let v = dilate ? 0 : 1;
      for (let k = -r; k <= r; k++) {
        const xx = x + k;
        if (xx < 0 || xx >= w) continue;
        v = pick(v, mask[y * w + xx]);
      }
      tmp[y * w + x] = v;
    }
  for (let y = 0; y < h; y++)
    for (let x = 0; x < w; x++) {
      let v = dilate ? 0 : 1;
      for (let k = -r; k <= r; k++) {
        const yy = y + k;
        if (yy < 0 || yy >= h) continue;
        v = pick(v, tmp[yy * w + x]);
      }
      out[y * w + x] = v;
    }
  return out;
}

/**
 * Head mask: low absolute threshold against a zero background, morphological
 * close, largest connected component, holes filled.
 *
 * Deliberately NOT Otsu -- see the docstring in preprocess.py. Otsu picks a
 * tissue-level threshold that fragments dim scans.
 */
export function headMask(g: Gray): Uint8Array {
  const { w, h } = g;
  const thr = Math.max(8, 0.1 * percentile(g.data, null, 99));
  const b = blur5(g);
  const raw = new Uint8Array(w * h);
  for (let i = 0; i < raw.length; i++) raw[i] = b.data[i] > thr ? 1 : 0;

  // Morphological close = dilate then erode with a 9x9 square (r = 4).
  const m = morph(morph(raw, w, h, 4, true), w, h, 4, false);

  // Largest connected component (iterative flood fill; no recursion depth risk).
  const label = new Int32Array(w * h).fill(-1);
  const stack = new Int32Array(w * h);
  let best = -1;
  let bestSize = 0;
  let cur = 0;
  for (let s = 0; s < m.length; s++) {
    if (!m[s] || label[s] >= 0) continue;
    let sp = 0;
    stack[sp++] = s;
    label[s] = cur;
    let size = 0;
    while (sp > 0) {
      const p = stack[--sp];
      size++;
      const px = p % w;
      const py = (p / w) | 0;
      for (let dy = -1; dy <= 1; dy++)
        for (let dx = -1; dx <= 1; dx++) {
          const nx = px + dx;
          const ny = py + dy;
          if (nx < 0 || ny < 0 || nx >= w || ny >= h) continue;
          const q = ny * w + nx;
          if (m[q] && label[q] < 0) {
            label[q] = cur;
            stack[sp++] = q;
          }
        }
    }
    if (size > bestSize) {
      bestSize = size;
      best = cur;
    }
    cur++;
  }

  const comp = new Uint8Array(w * h);
  if (best < 0) return comp.fill(1);
  for (let i = 0; i < comp.length; i++) comp[i] = label[i] === best ? 1 : 0;

  // Fill holes: flood the background inward from the border; anything still
  // unset is enclosed by the head and belongs to it (e.g. the ventricles).
  const outside = new Uint8Array(w * h);
  let sp = 0;
  const st = new Int32Array(w * h);
  for (let x = 0; x < w; x++) {
    for (const p of [x, (h - 1) * w + x]) if (!comp[p] && !outside[p]) { outside[p] = 1; st[sp++] = p; }
  }
  for (let y = 0; y < h; y++) {
    for (const p of [y * w, y * w + w - 1]) if (!comp[p] && !outside[p]) { outside[p] = 1; st[sp++] = p; }
  }
  while (sp > 0) {
    const p = st[--sp];
    const px = p % w;
    const py = (p / w) | 0;
    for (const [dx, dy] of [[1, 0], [-1, 0], [0, 1], [0, -1]]) {
      const nx = px + dx;
      const ny = py + dy;
      if (nx < 0 || ny < 0 || nx >= w || ny >= h) continue;
      const q = ny * w + nx;
      if (!comp[q] && !outside[q]) { outside[q] = 1; st[sp++] = q; }
    }
  }
  const filled = new Uint8Array(w * h);
  for (let i = 0; i < filled.length; i++) filled[i] = comp[i] || !outside[i] ? 1 : 0;
  return filled;
}

export interface Preprocessed {
  gray: Float32Array;   // size*size, 0..255 (for display)
  tensor: Float32Array; // 3*size*size, normalised (for the model)
  size: number;
}

export function preprocess(img: ImageData, size = 224, margin = 0.04, mean = 0.449, std = 0.226): Preprocessed {
  const g0 = toGray(img);

  // Un-squash: the raw files are 496x248, a 248x248 slice stretched 2x wide.
  const side = Math.max(g0.w, g0.h);
  const sq = g0.w === g0.h ? g0 : resize(g0, side, side);

  const mask = headMask(sq);

  // Square bounding box around the head, with margin, clamped to the frame.
  let y0 = sq.h, y1 = -1, x0 = sq.w, x1 = -1;
  for (let y = 0; y < sq.h; y++)
    for (let x = 0; x < sq.w; x++)
      if (mask[y * sq.w + x]) {
        if (y < y0) y0 = y;
        if (y > y1) y1 = y;
        if (x < x0) x0 = x;
        if (x > x1) x1 = x;
      }
  if (y1 < 0) { y0 = 0; x0 = 0; y1 = sq.h - 1; x1 = sq.w - 1; }
  y1 += 1; x1 += 1;

  const cy = (y0 + y1) / 2;
  const cx = (x0 + x1) / 2;
  const half = (Math.max(y1 - y0, x1 - x0) * (1 + 2 * margin)) / 2;
  const cy0 = Math.max(0, Math.round(cy - half));
  const cy1 = Math.min(sq.h, Math.round(cy + half));
  const cx0 = Math.max(0, Math.round(cx - half));
  const cx1 = Math.min(sq.w, Math.round(cx + half));

  const cw = cx1 - cx0;
  const ch = cy1 - cy0;
  const crop = new Float32Array(cw * ch);
  const cmask = new Uint8Array(cw * ch);
  for (let y = 0; y < ch; y++)
    for (let x = 0; x < cw; x++) {
      crop[y * cw + x] = sq.data[(cy0 + y) * sq.w + (cx0 + x)];
      cmask[y * cw + x] = mask[(cy0 + y) * sq.w + (cx0 + x)];
    }

  // Percentiles over head pixels only: over the whole frame they would mostly
  // measure how much black background there is.
  const lo = percentile(crop, cmask, 1);
  let hi = percentile(crop, cmask, 99);
  if (hi <= lo) hi = lo + 1;
  const normed = new Float32Array(cw * ch);
  for (let i = 0; i < crop.length; i++) {
    normed[i] = Math.min(1, Math.max(0, (crop[i] - lo) / (hi - lo))) * 255;
  }

  const final = resize({ data: normed, w: cw, h: ch }, size, size);
  const gray = new Float32Array(size * size);
  for (let i = 0; i < gray.length; i++) gray[i] = Math.round(final.data[i]);

  const tensor = new Float32Array(3 * size * size);
  for (let i = 0; i < size * size; i++) {
    const v = (gray[i] / 255 - mean) / std;
    tensor[i] = v;
    tensor[size * size + i] = v;
    tensor[2 * size * size + i] = v;
  }
  return { gray, tensor, size };
}

File: src/neurolink/data/preprocess.py
"""Preprocess raw OASIS JPGs into a single uint8 memmap.

What "regularize the images" actually means for this dataset:

1. RGB -> grayscale. The JPGs are stored 3-channel but the channels are
   identical; MRI is single-channel intensity data.
2. Un-squash. Files are 496x248 -- the true 248x248 axial slice stretched 2x
   horizontally. Resizing to square restores real anatomical proportions.
   Skipping this feeds the CNN systematically distorted brains.
3. Crop to the head's bounding box (square, with margin). This is the reel's
   "centering the pixels", done properly: it removes variation in head position
   and field-of-view padding so the network sees anatomy, not framing.
4. Intensity normalise using percentiles computed INSIDE the head mask. Taking
   percentiles over the whole frame would just measure how much black
   background there is.
5. Resize to 224x224 and store uint8.

The result is one contiguous memmap, so training epochs page in raw bytes
instead of decoding 86k JPEGs each time -- the difference between a feasible
overnight cross-validation and an infeasible one.
"""

from __future__ import annotations

import argparse
import multiprocessing as mp
import os
import shutil
from pathlib import Path

import cv2
import numpy as np
import pandas as pd
from tqdm import tqdm


def to_grayscale(img: np.ndarray) -> np.ndarray:
    if img.ndim == 3:
        return cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
    return img


def head_mask(gray: np.ndarray) -> np.ndarray:
    """Filled binary mask of the head.

    NOT Otsu. The background in these scans is exactly zero (all four corners
    measure 0.0), so the head/background boundary is trivial -- but Otsu instead
    picks a threshold that splits *tissue* (~50 on a 0-255 range). On a dim scan
    like OAS1_0287, whose median head intensity is 21, that threshold cuts
    through the brain itself, fragments the head into pieces, and the
    largest-component step then keeps only the brightest fragment. The bounding
    box lands mid-skull and the intensity normalisation, computed over that
    sliver, blows the image out.

    A low absolute threshold scaled to the image's own dynamic range separates
    head from a zero background robustly at any brightness. Holes are then filled
    so that dark interior structures -- the ventricles, which are the very thing
    we care about -- cannot punch the mask apart.
    """
    thr = max(8.0, 0.10 * float(np.percentile(gray, 99)))
    binm = (cv2.GaussianBlur(gray, (5, 5), 0) > thr).astype(np.uint8)
    binm = cv2.morphologyEx(binm, cv2.MORPH_CLOSE, np.ones((9, 9), np.uint8))

    n, labels, stats, _ = cv2.connectedComponentsWithStats(binm, connectivity=8)
    if n <= 1:
        return np.ones_like(gray, dtype=bool)
    largest = 1 + int(np.argmax(stats[1:, cv2.CC_STAT_AREA]))
    comp = (labels == largest).astype(np.uint8)

    # Fill interior holes: keep only outer contours and flood their interiors.
    contours, _ = cv2.findContours(comp, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
    filled = np.zeros_like(comp)
    cv2.drawContours(filled, contours, -1, 1, thickness=cv2.FILLED)
    return filled.astype(bool)


def square_bbox(mask: np.ndarray, margin: float, shape: tuple[int, int]) -> tuple[int, int, int, int]:
    """Square bounding box around the mask, clamped to the image.

    Square matters: a non-square crop resized to 224x224 would re-introduce
    exactly the anisotropic distortion step 2 just removed.
    """
    ys, xs = np.where(mask)
    if len(ys) == 0:
        h, w = shape
        return 0, 0, w, h
    y0, y1 = int(ys.min()), int(ys.max()) + 1
    x0, x1 = int(xs.min()), int(xs.max()) + 1

    cy, cx = (y0 + y1) / 2, (x0 + x1) / 2
    side = max(y1 - y0, x1 - x0) * (1 + 2 * margin)
    half = side / 2

    H, W = shape
    y0 = max(0, int(round(cy - half)));  y1 = min(H, int(round(cy + half)))
    x0 = max(0, int(round(cx - half)));  x1 = min(W, int(round(cx + half)))
    return x0, y0, x1, y1


def normalise(gray: np.ndarray, mask: np.ndarray, lo_p: float, hi_p: float) -> np.ndarray:
    """Percentile clip + rescale to [0, 255], using head pixels only."""
    vals = gray[mask] if mask.any() else gray.ravel()
    lo, hi = np.percentile(vals, [lo_p, hi_p])
    if hi <= lo:
        lo, hi = float(gray.min()), float(max(gray.max(), gray.min() + 1))
    out = (gray.astype(np.float32) - lo) / (hi - lo)
    return np.clip(out, 0, 1)


def process_one(
    path: Path, size: int, margin: float, clip: tuple[float, float]
) -> tuple[np.ndarray, np.ndarray] | None:
    """Returns (processed uint8 image, head mask at output resolution)."""
    raw = cv2.imread(str(path), cv2.IMREAD_UNCHANGED)
    if raw is None:
        return None
    gray = to_grayscale(raw)

    # Un-squash: 496x248 -> square, restoring true proportions.
    side = max(gray.shape)
    if gray.shape[0] != gray.shape[1]:
        gray = cv2.resize(gray, (side, side), interpolation=cv2.INTER_CUBIC)

    mask = head_mask(gray)
    x0, y0, x1, y1 = square_bbox(mask, margin, gray.shape)
    gray_c, mask_c = gray[y0:y1, x0:x1], mask[y0:y1, x0:x1]

    norm = normalise(gray_c, mask_c, *clip)
    out = cv2.resize(norm, (size, size), interpolation=cv2.INTER_AREA)
    out_mask = cv2.resize(mask_c.astype(np.uint8), (size, size), interpolation=cv2.INTER_NEAREST)
    return (out * 255).astype(np.uint8), out_mask.astype(bool)



def _worker(task: tuple[int, str, int, float, tuple[float, float]]):
    """Module-level so it is picklable by multiprocessing."""
    i, rel, size, margin, clip = task
    res = process_one(_WORKER_ROOT / rel, size, margin, clip)
    if res is None:
        return i, None, None
    return i, res[0], np.packbits(res[1].ravel())


_WORKER_ROOT: Path = Path(".")


def _init_worker(root: str) -> None:
    global _WORKER_ROOT
    _WORKER_ROOT = Path(root)
    cv2.setNumThreads(1)  # each worker is already a process; avoid thread thrash


def unpack_masks(packed: np.ndarray, size: int) -> np.ndarray:
    """Inverse of the bit-packing used by build_cache."""
    flat = np.unpackbits(packed, axis=-1, count=size * size)
    return flat.reshape(*packed.shape[:-1], size, size).astype(bool)


def build_cache(repo: Path, size: int, margin: float, clip: tuple[float, float]) -> None:
    manifest = pd.read_csv(repo / "cache" / "manifest.csv")
    n = len(manifest)

    # Head masks are stored BIT-PACKED. At uint8 they would cost as much as the
    # images themselves (4.3 GB); packed they cost 0.54 GB. They are needed for
    # the whole dataset, not just a sample, because the shortcut probe trains on
    # brain-removed images.
    img_gb = n * size * size / 1e9
    msk_gb = n * (size * size // 8) / 1e9
    free_gb = shutil.disk_usage(repo).free / 1e9
    print(f"Cache plan: images {img_gb:.2f} GB + packed masks {msk_gb:.2f} GB "
          f"= {img_gb + msk_gb:.2f} GB;  free: {free_gb:.1f} GB")
    if free_gb < img_gb + msk_gb + 3.0:
        raise SystemExit(
            f"Refusing to start: only {free_gb:.1f} GB free, need "
            f"{img_gb + msk_gb + 3.0:.1f} GB including headroom. Re-run with a smaller --size."
        )

    assert size * size % 8 == 0, "size^2 must be divisible by 8 for bit packing"
    img_path = repo / "cache" / f"slices_{size}.npy"
    msk_path = repo / "cache" / f"masks_{size}_packed.npy"
    images = np.lib.format.open_memmap(img_path, mode="w+", dtype=np.uint8, shape=(n, size, size))
    masks = np.lib.format.open_memmap(
        msk_path, mode="w+", dtype=np.uint8, shape=(n, size * size // 8)
    )

    # Each image is independent, so this parallelises perfectly. Single-process
    # throughput is ~29 img/s (a 49-minute pass); a worker pool cuts that to a
    # few minutes on this machine.
    n_proc = max(1, min(os.cpu_count() or 4, 10))
    tasks = [(i, rel, size, margin, clip) for i, rel in enumerate(manifest["path"])]
    print(f"preprocessing with {n_proc} worker processes")

    failed = []
    with mp.Pool(n_proc, initializer=_init_worker, initargs=(str(repo / "data"),)) as pool:
        for i, img, packed in tqdm(
            pool.imap_unordered(_worker, tasks, chunksize=64),
            total=n, desc="preprocess", unit="img"
        ):
            if img is None:
                failed.append(manifest["path"].iloc[i])
                continue
            images[i] = img
            masks[i] = packed

    images.flush(); masks.flush()
    if failed:
        print(f"WARNING: {len(failed)} images failed to decode: {failed[:5]}")
        (repo / "cache" / "failed.txt").write_text("\n".join(failed))
    print(f"Wrote {img_path.name} ({img_gb:.2f} GB) and {msk_path.name} ({msk_gb:.2f} GB)")


if __name__ == "__main__":
    ap = argparse.ArgumentParser()
    ap.add_argument("--size", type=int, default=224)
    ap.add_argument("--margin", type=float, default=0.04)
    ap.add_argument("--probe", type=int, default=0, help="render N sanity samples instead of building")
    args = ap.parse_args()

    repo = Path(__file__).resolve().parents[3]
    clip = (1.0, 99.0)

    if args.probe:
        from .probe import render_probe
        render_probe(repo, args.size, args.margin, clip, args.probe)
    else:
        build_cache(repo, args.size, args.margin, clip)
